src/attribute.cpp

```diff
@@ -188,93 +222,51 @@ std::vector<uint8_t> Bytes::toVector(const std::string& /*secret*/, const std::a
     return attribute;
 }
 
-std::string String::value() const
+ChapPassword::ChapPassword(uint8_t type, const uint8_t* data, size_t size)
+    : Attribute(type),
+      m_value(size - 1)
 {
-    return m_value;
-}
+    if (size != 17)
+        throw std::runtime_error("Invalid CHAP_PASSWORD attribute size. Should be 17, actual size is " + std::to_string(size));
 
-std::string Integer::value() const
-{
-    return std::to_string(m_value);
-}
+    m_chapId = data[0];
 
-std::string IpAddress::value() const
-{
-    return std::to_string(m_value[0]) + "." + std::to_string(m_value[1]) + "." + std::to_string(m_value[2]) + "." + std::to_string(m_value[3]);
+    for (size_t i = 0; i < size - 1; ++i)
+        m_value[i] = data[i + 1];
 }
 
-std::string Encrypted::value() const
+ChapPassword::ChapPassword(uint8_t type, uint8_t chapId, const std::vector<uint8_t>& chapValue)
+    : Attribute(type),
+      m_chapId(chapId),
+      m_value(chapValue)
 {
-    return m_value;
 }
 
-std::string Bytes::value() const
+std::string ChapPassword::toString() const
 {
     std::string value;
 
     for (const auto& b : m_value)
         value += byteToHex(b);
 
-    return value;
+    return std::to_string(m_chapId) + value;
```

Comment:
This way CHAP id will be hard to distinguish from the value. Let's separate them by something. Spacebar, coma, etc.